File: Subin/week2/할인행사.py
def solution(want, number, discount):
    answer = 0
    buy = {}
    for i in range(len(want)):
        buy[want[i]] = number[i] #딕셔너리로 구매 리스트 만듦
    all = sum(number) #총 사야하는 물건 수
    #10일씩 몇 번 묶을것인가
    for x in range(len(discount)-9):
        check = buy.copy() #수량 비교 위한 사전
        check_num = 0 #다 구매했는지 확인하기 위한 변수
        
        for y in range(10):
            if discount[x+y] in check and check[discount[x+y]]>0:
                check[discount[x+y]]-=1
                check_num += 1
                
        if check_num == all: #다 구매됐을때만 answer++
            answer +=1

    return answer

File: nahyun/week2/131127. 할인 행사/할인 행사.java
import java.util.*;
import java.io.*;

class Solution {
    public static int solution(String[] want, int[] number, String[] discount) {
        Map<String, Integer> map = new HashMap<>();
        int disLastIndex = discount.length - 1;
        // 초기 10개의 항목을 map에 추가
        for (int i = 0; i < 10; i++) {
            if (i > disLastIndex) break;
            String str = discount[i];
            map.put(str, map.getOrDefault(str, 0) + 1);
        }

        int count = 0;
        for (int i = 0; i + 9 <= disLastIndex; i++) {
            if (check(want, number, map)) count++;
            
            // map 업데이트
            if (i + 10 <= disLastIndex) {
                String removeKey = discount[i];
                String addKey = discount[i + 10];

                // 전 시작 값을 제거
                map.put(removeKey, map.get(removeKey) - 1);
                if (map.get(removeKey) == 0) {
                    map.remove(removeKey);
                }

                // 새로운 끝 값을 추가
                map.put(addKey, map.getOrDefault(addKey, 0) + 1);
            }
        }
        return count;
    }

    public static boolean check(String[] want, int[] number, Map<String, Integer> map) {
        for (int i = 0; i < want.length; i++) {
            String key = want[i];
            if (!map.containsKey(key) || map.get(key) < number[i]) {
                return false;
            }
        }
        return true;
    }
}
